merkle_tree_wasm/src/merkle_tree_wasm.rs

```diff
@@ -0,0 +1,213 @@
+
+        let depth = (padding_len as f64).log2() as usize;
+
+        self.depth = Some(depth);
+
+        self.is_tree_ready = true;
+
+        // Calculate the root
+        let root = self.calculate_root();
+        self.root = Some(root);
+    }
+
+    pub fn calculate_root(&mut self) -> F {
+        if !self.is_tree_ready {
+            panic!("MerkleTree: Tree is not ready.");
+        }
+
+        self.layers.push(self.leaves.clone());
+        let mut current_layer = self.layers[0].clone();
+
+        for _ in 0..self.depth.unwrap() {
+            let layer_above = current_layer
+                .par_chunks(self.arity())
+                .map(|nodes| {
+                    let mut poseidon = self.poseidon.clone();
+                    Self::hash(&mut poseidon, nodes)
+                })
+                .collect::<Vec<F>>();
+
+            self.layers.push(layer_above.clone());
+            current_layer = layer_above;
+        }
+
+        assert_eq!(current_layer.len(), 1);
+
+        current_layer[0]
+    }
+
+    fn arity(&self) -> usize {
+        WIDTH - 1
+    }
+
+    pub fn create_proof(&self, leaf: F) -> MerkleProof<F> {
+        if !self.is_tree_ready {
+            panic!("MerkleTree: Tree is not ready.");
+        }
+
+        let mut siblings = vec![];
+        let mut path_indices = vec![];
+
+        let mut current_layer = &self.layers[0];
+
+        let mut leaf_index = self
+            .leaves
+            .iter()
+            .position(|&x| x == leaf)
+            .expect("MerkleProof: Leaf is not found");
+
+        for i in 0..self.depth.unwrap() {
+            let sibling_index = if leaf_index % 2 == 0 {
+                leaf_index + 1
+            } else {
+                leaf_index - 1
+            };
+
+            let sibling = current_layer[sibling_index];
+            siblings.push(sibling);
+            path_indices.push(leaf_index & 1);
+
+            leaf_index /= 2;
+            current_layer = &self.layers[i + 1];
+        }
+
+        // TODO: encode the output of the merkle_proof into big-endian bytes
+
+        MerkleProof {
+            leaf,
+            siblings,
+            path_indices,
+            root: self.root.unwrap(),
+        }
+    }
+
+    pub fn verify_proof(&mut self, root: F, proof: &MerkleProof<F>) -> bool {
+        let mut node = proof.leaf;
+        for (sibling, node_index) in proof.siblings.iter().zip(proof.path_indices.iter()) {
+            let is_node_index_even = node_index & 1 == 0;
+
+            let nodes = if is_node_index_even {
+                [node, *sibling]
+            } else {
+                [*sibling, node]
+            };
+
+            node = Self::hash(&mut self.poseidon, &nodes);
+        }
+
+        node == root
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ark_std::{end_timer, start_timer};
+    use poseidon::constants::secp256k1_w3;
+
+    type F = ark_secp256k1::Fq;
+
+    fn test_tree() {
```

Comment:
```suggestion
   #[test]
    fn test_tree() {
```

Otherwise test won't run